sysdump/sysdump.go

```diff
@@ -0,0 +1,815 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sysdump
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium-cli/internal/utils"
+	"github.com/cilium/workerpool"
+	"github.com/hashicorp/go-multierror"
+	archiver "github.com/mholt/archiver/v3"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+type SysdumpOptions struct {
+	CiliumLabelSelector         string
+	CiliumNamespace             string
+	CiliumOperatorLabelSelector string
+	CiliumOperatorNamespace     string
+	Debug                       bool
+	HubbleLabelSelector         string
+	HubbleNamespace             string
+	HubbleRelayLabelSelector    string
+	HubbleRelayNamespace        string
+	HubbleUILabelSelector       string
+	HubbleUINamespace           string
+	LargeSysdumpAbortTimeout    time.Duration
+	LargeSysdumpThreshold       int32
+	LogsLimitBytes              int64
+	LogsSinceTime               time.Duration
+	NodeList                    string
+	OutputFileName              string
+	Quick                       bool
+	RESTClientGetter            genericclioptions.RESTClientGetter
+	WorkerCount                 int32
+	Writer                      io.Writer
+}
+
+type Sysdump interface {
```

Comment:
not sure why this is an interface. Are we planning to have multiple sysdump implementations? (maybe they already exist and I haven't seen it yet)

Reply:
We're using this pattern in [some](https://github.com/cilium/cilium-cli/blob/master/hubble/hubble.go#L42) [places](https://github.com/cilium/cilium-cli/blob/master/connectivity/check/check.go#L173), and I think that even if it doesn't add anything for now, it doesn't hurt either. WDYT?

Reply:
I would only create the interface if it's strictly required, e.g. to implement tests.